hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/AlterHoodieTableAddPartitionCommand.scala

```diff
@@ -0,0 +1,88 @@
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.catalog.CatalogTypes.TablePartitionSpec
+import org.apache.spark.sql.catalyst.catalog.{CatalogTablePartition, HoodieCatalogTable}
+import org.apache.spark.sql.execution.command.DDLUtils
+import org.apache.spark.sql.hudi.HoodieSqlCommonUtils.{makePartitionPath, normalizePartitionSpec}
+
+case class AlterHoodieTableAddPartitionCommand(
+   tableIdentifier: TableIdentifier,
+   partitionSpecsAndLocs: Seq[(TablePartitionSpec, Option[String])],
+   ifNotExists: Boolean)
+  extends HoodieLeafRunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val fullTableName = s"${tableIdentifier.database}.${tableIdentifier.table}"
+    logInfo(s"start execute alter table add partition command for $fullTableName")
+
+    val hoodieCatalogTable = HoodieCatalogTable(sparkSession, tableIdentifier)
+
+    if (!hoodieCatalogTable.isPartitionedTable) {
+      throw new AnalysisException(s"$fullTableName is a non-partitioned table that is not allowed to add partition")
+    }
+
+    val catalog = sparkSession.sessionState.catalog
+    val table = hoodieCatalogTable.table
+    DDLUtils.verifyAlterTableType(catalog, table, isView = false)
+
+    val normalizedSpecs: Seq[Map[String, String]] = partitionSpecsAndLocs.map { case (spec, location) =>
+      if (location.isDefined) {
+        throw new AnalysisException(s"Hoodie table does not support specify partition location explicitly")
+      }
+      normalizePartitionSpec(
+        spec,
+        hoodieCatalogTable.partitionFields,
+        hoodieCatalogTable.tableName,
+        sparkSession.sessionState.conf.resolver)
+    }
+
+    val basePath = new Path(hoodieCatalogTable.tableLocation)
+    val fileSystem = hoodieCatalogTable.metaClient.getFs
+    val instantTime = HoodieActiveTimeline.createNewInstantTime
+    val format = hoodieCatalogTable.tableConfig.getPartitionMetafileFormat
+    val (partitionMetadata, parts) = normalizedSpecs.map { spec =>
+      val partitionPath = makePartitionPath(hoodieCatalogTable, spec)
+      val fullPartitionPath: Path = FSUtils.getPartitionPath(basePath, partitionPath)
+      val metadata = if (HoodiePartitionMetadata.hasPartitionMetadata(fileSystem, fullPartitionPath)) {
+        if (!ifNotExists) {
+          throw new AnalysisException(s"Partition metadata already exists for path: $fullPartitionPath")
+        }
+        None
+      } else Some(new HoodiePartitionMetadata(fileSystem, instantTime, basePath, fullPartitionPath, format))
+      (metadata, CatalogTablePartition(spec, table.storage.copy(locationUri = Some(fullPartitionPath.toUri))))
+    }.unzip
+    partitionMetadata.flatten.foreach(_.trySave(0))
+
+    // Sync new partitions in batch, enable ignoreIfExists to avoid sync failure.
+    val batchSize = sparkSession.sparkContext.conf.getInt("spark.sql.addPartitionInBatch.size", 100)
+    parts.toIterator.grouped(batchSize).foreach { batch =>
```

Comment:
Should we try catch here? Looks `CreateHoodieTableCommand` catch exception if syncing to HMS occurs error.

Reply:
ping @danny0405 any thoughts for this? I see some commands catch the exception and only print warning(like `CreateHoodieTableCommand`), and some commands throw the exception out(like `DropHoodieTableCommand`, `CreateHoodieTableAsSelectCommand`)
Looks we don't have any standards whether throw exceptions if syncing to HMS occurs error.

Reply:
I have no preference, if we throw exception for hive sync, will the table created be cleaned ?

Reply:
It makes sense to add a warning log, otherwise the exception will be displayed in client while we succeed add partition folder actually.